python/phase-2.md

```diff
@@ -0,0 +1,830 @@
+            # Store greeting (this is typed)
+            self._current_greeting = await workflow.execute_local_activity(
+                create_greeting_activity,
+                self._salutation,
+                name,
+                start_to_close_timeout=timedelta(seconds=5),
+            )
+            workflow.logger.debug("Greeting set to %s", self._current_greeting)
+            
+            # Wait for salutation update or complete signal (this can be
+            # cancelled)
+            await asyncio.wait(
+                [self._salutation_update.wait(), self._complete.wait()],
+                return_when=asyncio.FIRST_COMPLETED,
+            )
+            if self._salutation_update.is_set():
+                self._salutation_update.clear()
+            elif self._complete.is_set():
+                return self._current_greeting
+
+    @workflow.signal
+    async def update_salutation(self, salutation: str) -> None:
+        self._salutation = salutation
+        self._salutation_update.set()
+
+    @workflow.signal
+    async def complete_with_greeting(self) -> None:
+        self._complete.set()
+
+    @workflow.query
+    async def current_greeting(self) -> str:
+        return self._current_greeting
+
+def test_workflow(client: Client) -> None:
+    # Start the workflow
+    handle = await client.start_workflow(GreetingWorkflow.run, "John")
+
+    # Confirm the current greeting
+    assert "Hello, John!" == await handle.query(GreetingWorkflow.current_greeting)
+
+    # Change the salutation part
+    await handle.signal(GreetingWorkflow.update_salutation, "Hola")
+
+    # Now tell it to complete with the new greeting
+    await handle.signal(GreetingWorkflow.complete_with_greeting)
+
+    # Check result
+    assert "Hola, John!" == await handle.result()
+
+def create_worker(client: Client) -> Worker:
+    return Worker(
+      client=client,
+      activities=[create_greeting_activity],
+      workflows=[GreetingWorkflow],
+    )
+```
+
+Notes:
+
+* Workflows are classes
+  * Consideration was given to making workflows be functions (in addition to classes or having the class equivalent sit
+    on top) with several possibilities for defining signals and queries, but all were too ugly/complicated
+* `@activity.defn`
+  * Must be on function and accepts an optional activity `name` which defaults to function `__name__`
+  * This was originally not wanted in the activity impl (phase 1) due to complaints of the name not being obvious
+  * We need to define the activity name on the activity for execution by the workflow
+  * Is it ok to now make this required? Is it ok to now make the name implicit? Is it ok to now use a list instead of a
+    dict to specify activities for the worker?
+* `@workflow.defn`
+  * Must be on class and accepts an optional workflow `name` which defaults to class `__name__`
+  * This is required to clearly mark the class as a workflow
+  * No base classes can have this value
+* `@workflow.run`
```

Comment:
You might want to call this `execute` to match the client API

Reply:
Why not `@workflow.start` which matches the client API? I considered both and figured this was the shortest and clearest to define what it is. I also considered `@workflow.entry` but I don't think that makes it as clear that the workflow only lasts for the life of the function.

Reply:
`execute` matches the API, not `start`, the client `execute_workflow` method copies this signature.

Reply:
I don't mind changing to `@workflow.execute` if others agree too.